wallet: finalize `PsbtManager` implementation

yyforyongyu opened the pull request:
This PR fixes #1035 - the old design has evolved here, detailed below.

## Initial State: Fragmented and Rigid Methods

In the original monolithic `wallet.Interface`, PSBT-related methods (`FundPsbt`, `SignPsbt`, `FinalizePsbt`) were mixed in with general transaction and account management methods. This had several drawbacks:

1.  **Hidden Workflow**: The PSBT lifecycle (`Fund` -> `Decorate` -> `Sign` -> `Combine` -> `Finalize`) is a distinct, multi-step process. Scattering these methods obscured this workflow, making it harder for developers to understand the correct sequence of operations.
2.  **Rigid Signatures**: The original method signatures relied on long lists of primitive arguments (e.g., `FundPsbt` taking explicit `minConfs`, `feeRate`, `account`, `keyScope`, `strategy`). This made the API brittle; adding a new funding option (like a specific coin selection strategy or a change output policy) would require breaking changes.
3.  **Missing Functionality**: The wallet lacked a dedicated `CombinePsbt` method, forcing users to rely on external tools or ad-hoc merging logic for collaborative signing flows (like Multisig or CoinJoin).

## The Design Evolution: A Cohesive, Extensible Workflow

The `PsbtManager` was designed to solve these issues by grouping all PSBT operations into a single, cohesive interface and adopting extensible parameter objects.

### Insight 1: Workflow Cohesion and Discovery

By extracting `PsbtManager` into its own interface, we elevate PSBTs to a first-class citizen in the wallet API.

-   **Explicit Workflow**: Grouping `Fund`, `Decorate`, `Sign`, `Combine`, and `Finalize` together makes the intended workflow self-evident.
-   **Separation of Concerns**: This prevents "interface bloat" in `TxPublisher` or `Signer`. While PSBTs involve signing, the high-level coordination of a PSBT packet is conceptually distinct from the low-level cryptographic operations of the `Signer`.

### Insight 2: Extensibility via Parameter Objects (`FundIntent`)

Instead of passing numerous arguments to `FundPsbt`, the new design introduces the `FundIntent` struct.

-   **Declarative Intent**: `FundIntent` allows the caller to describe *what* they want (inputs, outputs, fee rate, change source) rather than just passing values.
-   **Flexible Input Strategy**: The `Intent` allows for a `Policy` (automatic coin selection) or manual `Inputs` (coin control), providing flexibility without changing the method signature.
-   **Future-Proofing**: New funding options can be added to `FundIntent` (e.g., specific coin selection algorithms, RBF preferences) without breaking existing code. The same pattern is applied to `SignPsbt` via `SignPsbtParams`.

### Insight 3: The "Combiner" Role

The addition of `CombinePsbt` fills a critical gap for collaborative workflows.

-   **BIP-174 Compliance**: It implements the formal "Combiner" role defined in BIP-174, allowing multiple partial signatures for the same inputs to be merged into a single packet.
-   **Stateless Merging**: The implementation is designed to be safe and robust, validating that all packets refer to the same transaction before merging. It handles the complexity of deduplicating signatures, scripts, and derivation paths, freeing the client from error-prone manual merging.

## The Final API: A Complete PSBT Toolkit

The final design provides a robust, extensible, and complete toolkit for managing the PSBT lifecycle.

```go
// PsbtManager provides a cohesive, high-level interface for creating and
// managing Partially Signed Bitcoin Transactions (PSBTs).
type PsbtManager interface {
	// DecorateInputs enriches a PSBT's inputs with UTXO and derivation
	// information known to the wallet.
	DecorateInputs(ctx context.Context, packet *psbt.Packet,
		skipUnknown bool) (*psbt.Packet, error)

	// FundPsbt performs coin selection and adds the selected inputs (and a
	// change output, if necessary) to the PSBT.
	FundPsbt(ctx context.Context, intent *FundIntent) (*psbt.Packet,
		int32, error)

	// SignPsbt adds partial signatures to the PSBT for all inputs
	// controlled by the wallet.
	SignPsbt(ctx context.Context, params *SignPsbtParams) (
		*SignPsbtResult, error)

	// FinalizePsbt attempts to finalize the PSBT, transitioning it from a
	// partially signed state to a complete, network-ready transaction.
	FinalizePsbt(ctx context.Context, packet *psbt.Packet) error

	// CombinePsbt acts as the "Combiner" role in BIP 174, merging multiple
	// Partially Signed Bitcoin Transactions (PSBTs) into a single packet.
	CombinePsbt(ctx context.Context, psbts ...*psbt.Packet) (
		*psbt.Packet, error)
}
```

### Why This Design is Superior:

1.  **Workflow-Centric**: The interface clearly guides the developer through the lifecycle of a PSBT.
2.  **Extensible**: Parameter objects like `FundIntent` ensure the API can evolve to support advanced features (like custom coin selection strategies) without breaking changes.
3.  **Collaborative-Ready**: The inclusion of `CombinePsbt` and the support for external input tweakers in `SignPsbtParams` make this interface fully capable of handling complex, multi-party signing scenarios.

yyforyongyu committed:
wallet: implement `CombinePsbt`

yyforyongyu committed:
wallet: remove file `psbt.go`

We now move all helper methods into `psbt_manager.go`.

yyforyongyu committed:
wallet: patch unit tests for existing helper methods

Add dedicated unit tests for these old methods.

yyforyongyu committed:
wallet: add PSBT workflow docs

yyforyongyu committed:
wallet: implement `FinalizePsbt`

The new FinalizePsbt implementation diverges from the deprecated method
by decoupling signing logic from account scoping. Instead of requiring
an account number and checking for watch-only status explicitly, it
iterates over all inputs and attempts to sign them using the modern
Signer interface (ComputeUnlockingScript).

Advantages:
1. Robustness: It attempts to sign any input for which the wallet possesses
   the private key, supporting mixed-account transactions naturally.
2. Safety: Defensive checks in fetchPsbtUtxo and FinalizePsbt prevent panics
   on malformed PSBTs (missing UTXO info), logging warnings instead.
3. Multi-party Support: Inputs that cannot be signed (e.g., belonging to
   other parties) are skipped gracefully, allowing MaybeFinalizeAll to
   complete the transaction using existing partial signatures if present.

This approach simplifies the API (no account args) and aligns with the
PsbtManager's goal of being a general-purpose PSBT handler.

yyforyongyu committed:
wallet: consolidate PSBT errors and refactor manager

This commit consolidates various specific PSBT errors into broader
categories (ErrNilArguments, ErrInvalidBip32Path, ErrIndexOutOfBounds,
ErrPsbtMergeConflict) to reduce API surface area. It also refactors
fetchPsbtUtxo to return errors instead of logging, ensuring better
failure handling in SignPsbt and FinalizePsbt. Unit tests and linting
issues (err113, lll) have been addressed.

mohamedawnallah commented on wallet/psbt_manager.go:
```diff
@@ -1607,3 +1544,729 @@ func (w *Wallet) signBip32PsbtInput(ctx context.Context, packet *psbt.Packet,
+
+	// Post-merge Validation: Ensure the resulting packet is structurally
+	// sound (e.g. has necessary UTXO info). This acts as a final sanity
+	// check.
```
👍

mohamedawnallah commented on wallet/psbt_manager_test.go:
```diff
@@ -3277,15 +3277,1164 @@ func TestSignBip32PsbtInputErrors(t *testing.T) {
-		0, 0,
+// TestPatchInputWithScript tests that patchInputWithScript correctly updates
+// the PSBT input with the provided witness and/or sigScript.
+func TestPatchInputWithScript(t *testing.T) {
```
https://github.com/mohamedawnallah/btcwallet/pull/41#discussion_r2624447379

yyforyongyu replied:
fixed

mohamedawnallah commented on wallet/psbt_manager.go:
```diff
@@ -1607,3 +1544,729 @@ func (w *Wallet) signBip32PsbtInput(ctx context.Context, packet *psbt.Packet,
+	dest.TaprootScriptSpendSig = append(
+		dest.TaprootScriptSpendSig, src.TaprootScriptSpendSig...,
+	)
+}
```
https://github.com/mohamedawnallah/btcwallet/pull/41#discussion_r2624462841

yyforyongyu replied:
yeah right - should be a deduplicator than a merger

mohamedawnallah commented on wallet/psbt_manager_test.go:
```diff
@@ -3277,15 +3277,1164 @@ func TestSignBip32PsbtInputErrors(t *testing.T) {
 	}}
 	err = w.signBip32PsbtInput(t.Context(), packet, 0, nil, nil)
-	require.ErrorIs(t, err, ErrInvalidBip32PathLength)
+	require.ErrorIs(t, err, ErrInvalidBip32Path)
```
https://github.com/mohamedawnallah/btcwallet/pull/41#discussion_r2624467664

yyforyongyu replied:
Added back

mohamedawnallah reviewed (approved):
Gemini seems provided valid comments

GustavoStingelin commented on docs/developer/psbt_workflows.md:
````diff
@@ -0,0 +1,295 @@
+    ```
+
+2.  **Fund the PSBT:** Alice's wallet performs coin selection to add inputs and a
+    change output.
+
+    ```go
+    fundIntent := &wallet.FundIntent{
+        Packet:      packet,
+            Policy: &wallet.InputsPolicy{
+                Source: &wallet.ScopedAccount{
+                    AccountName: "default",
+                    KeyScope:    waddrmgr.KeyScopeBIP0086,
+                },
+                MinConfs: 1,
+            },
+            FeeRate: btcunit.NewSatPerKVByte(1000), // e.g., 1 sat/vb
+        }
+            fundedPacket, _, err := aliceWallet.FundPsbt(ctx, fundIntent)
+    ```
````
Looks like we have some indentation problem here

```suggestion
    ```go
    fundIntent := &wallet.FundIntent{
        Packet:      packet,
        Policy: &wallet.InputsPolicy{
                Source: &wallet.ScopedAccount{
                    AccountName: "default",
                    KeyScope:    waddrmgr.KeyScopeBIP0086,
                },
                MinConfs: 1,
            },
        FeeRate: btcunit.NewSatPerKVByte(1000), // e.g., 1 sat/vb
    }
    fundedPacket, _, err := aliceWallet.FundPsbt(ctx, fundIntent)
    ```
```

yyforyongyu replied:
Updated - also moved the doc under /docs/developer as it feels more fit.

GustavoStingelin commented on wallet/psbt_manager.go:
```diff
@@ -1607,3 +1544,729 @@ func (w *Wallet) signBip32PsbtInput(ctx context.Context, packet *psbt.Packet,
+		HashType:   pInput.SighashType,
+	}
+
+	unlockingScript, err := w.ComputeUnlockingScript(ctx, params)
+	if err != nil {
+		// If we can't generate the script (e.g. we don't own the key,
+		// or it's a type we don't support yet, or the account is
+		// watch-only), we just skip this input and let the finalizer
+		// try to use any existing partial signatures.
+		log.Debugf("Could not compute unlocking script for "+
+			"input %d: %v", idx, err)
+
+		return nil
+	}
```
Maybe we should check if the returned error is `ErrUnableToExtractAddress`? And if yes, throw it up?

yyforyongyu replied:
Yeah good point - I almost finished adding this check, then I realized the `FinalizePsbt` is just gives best effort - it tries to sign as much as it can, and skips otherwise. Here we may get this `ErrUnableToExtractAddress`, while it could be some customized script that is used and can be signed by another wallet?

I think in the end, it's about whether the tx can be finalized with valid sigs, and here, we just try to sign our inputs.

GustavoStingelin replied:
Hmm, that makes sense.

GustavoStingelin commented on wallet/psbt_manager.go:
Q: Any chance to get an `index out of range` here?

yyforyongyu replied:
Good call - also realized we can just use `fetchPsbtUtxo`, updated

GustavoStingelin commented on wallet/psbt_manager.go:
```diff
@@ -1607,3 +1544,729 @@ func (w *Wallet) signBip32PsbtInput(ctx context.Context, packet *psbt.Packet,
+		for _, dSig := range dest {
+			if bytes.Equal(dSig.PubKey, sig.PubKey) {
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			dest = append(dest, sig)
+		}
+	}
+
+	return dest
```
nit: this logic could be replaced by https://pkg.go.dev/slices#example-CompactFunc

Also for the other deduplicators.

yyforyongyu replied:
TIL - but we need to use `slices.ContainsFunc` instead tho, added a commit in the end.

GustavoStingelin replied:
cool

GustavoStingelin reviewed (changes requested):
Looking good, left a few comments.

yyforyongyu commented on wallet/psbt_manager.go:
```diff
@@ -1607,3 +1544,729 @@ func (w *Wallet) signBip32PsbtInput(ctx context.Context, packet *psbt.Packet,
+
+	// Fetch the UTXO for this input.
+	utxo, err := fetchPsbtUtxo(packet, idx)
+	if err != nil {
```
Note that previously we returned the error, but this is wrong - we should just skip this input.

yyforyongyu committed:
wallet: use `slices.ContainsFunc`

yyforyongyu committed:
wallet: return error from `PsbtPrevOutputFetcher`

yyforyongyu committed:
wallet: add deduplicateTaprootScriptSpendSigs and its test

This commit introduces a new helper function, deduplicateTaprootScriptSpendSigs, to correctly handle deduplication of Taproot script spend signatures when merging PSBTs. This aligns with BIP-174 combiner role and ensures consistency with other deduplication methods. A comprehensive unit test, TestDeduplicateTaprootScriptSpendSigs, has been added to verify its behavior.

GustavoStingelin reviewed (approved):
LGTM! 🚀 🚀 🚀
